design-proposals/e2e-framework.md

```diff
@@ -0,0 +1,92 @@
+Framework for E2E Tests
+=
+
+# Overview
+This proposal is suggesting the introduction of a `Framework` to support common operations in e2e tests.
+
+# Motivation
+
+Kubevirt has an impressive amount of e2e test, covering a large amount scenarios and features.
+
+Two paradigms can be used to make each test run on a clean environment:
+- restore a clean situation before the execution of each test.
+- make each test leave a clean situation.
+
+Currently, in kubevirt the first option is most commonly used, but specific cleanups are also
+performed after test execution. This mix can create leftovers and sometimes makes it difficult
+to understand where a problem may have occurred.
+
+It is also a symptom of a kind of non-independence of the tests from each other:
+test A does not care about cleaning the cluster, so much as it knows that there will be
+the setup of test B will do it.
+
+Last but not least, each test file must specify the cleanup and setup function on its own, and in most cases
+is something like this:
+```go
+BeforeEach(func() {
+  tests.BeforeTestCleanup()
+  
+  virtClient, err = kubecli.GetKubevirtClient()
+  util.PanicOnError(err)
+  //other test specific setup
+})
+```
+
+The `Framework`  can help resolve these issues by supporting tests in setting up and cleaning up the environment
+in which they will operate. This will provide a separation of duties in the various phases of a test.
+This will make it easier to script the tests since you will not have to worry about setup and cleanup
+of the environment. It will also define a test architecture that is easier for new members to understand.
+
+## Goals
+
+- Support common operations.
+- Increase e2e test stabilization.
+- Minimizing the possibility of leftovers after a test run.
+
+## Non Goals
+- Refactoring all the e2e tests at once
+
+# Proposal
+
+## Definition of Personas
+
+- E2E test authors.
+
+## User Stories
+
+As an e2e test author:
+
+I want to create a new e2e test file ensuring that there will be no leftover that may influence subsequent testing.
+
+## Repos
+Kubevirt/kubevirt
+
+## Design
```

Comment:
What about separating tests to different namespaces? This would ensure that even if the environment is not cleaned properly, other tests are less likely to be affected.

It probably doesn't make sense to create a different namespace for each test (or is it? :o), but maybe a namespace per 10 tests (or some other constant)? Maybe per test-suite?

Reply:
If that's possible from a performance/scalability perspective, I don't see a reason why not! :smiley:

Reply:
Note CDIs framework does exactly this, create a new namespace per test. We don't wait for the namespaces to be deleted before creating a new one. We delete the namespace and then move on without waiting.

Reply:
Thanks @awels, this could be an idea, since my tests show that deleting a namespace requires around 5 secs.